mpas_analysis/shared/timekeeping/MpasRelativeDelta.py

```diff
@@ -0,0 +1,131 @@
+import datetime
+from dateutil.relativedelta import relativedelta
+import calendar
+
+
+class MpasRelativeDelta(relativedelta):
+    """
+    MpasRelativeDelta is a subclass of dateutil.relativedelta for relative time
+    intervals with different MPAS calendars.
+
+    Only relative intervals (years, months, etc.) are supported and not the
+    absolute date specifications (year, month, etc.).  Addition/subtraction
+    of datetime.datetime objects (but not other MpasRelativeDelta,
+    datetime.timedelta or other related objects) is supported.
+
+    Author
+    ------
+    Xylar Asay-Davis
+
+    Last Modified
+    -------------
+    02/09/2017
+    """
+
+    def __init__(self, dt1=None, dt2=None, years=0, months=0, days=0,
+                 hours=0, minutes=0, seconds=0, calendar='gregorian'):
+        if calendar not in ['gregorian', 'gregorian_noleap']:
+            raise ValueError('Unsupported MPAs calendar {}'.format(calendar))
+        self.calendar = calendar
+        super(MpasRelativeDelta, self).__init__(dt1=dt1, dt2=dt2, years=years,
+                                                months=months, days=days,
+                                                hours=hours, minutes=minutes,
+                                                seconds=seconds)
+
+    def __add__(self, other):
+        if not isinstance(other, datetime.datetime):
+            return NotImplemented
+
+        year = other.year+self.years
+
+        month = other.month
+        if self.months != 0:
+            assert 1 <= abs(self.months) <= 12
+            month += self.months
+            if month > 12:
+                year += 1
+                month -= 12
+            elif month < 1:
+                year -= 1
+                month += 12
+
+        if self.calendar == 'gregorian':
+            daysInMonth = calendar.monthrange(year, month)[1]
+        elif self.calendar == 'gregorian_noleap':
+            # use year 0001, which is not a leapyear
+            daysInMonth = calendar.monthrange(1, month)[1]
+
+        day = min(daysInMonth, other.day)
+        repl = {"year": year, "month": month, "day": day}
+
+        days = self.days
+        if self.calendar == 'gregorian_noleap' and calendar.isleap(year):
+            if month == 2 and day+days >= 29:
+                # skip forward over the leap day
+                days += 1
+            elif month == 3 and day+days <= 0:
+                # skip backward over the leap day
+                days -= 1
+
+        return (other.replace(**repl) +
+                datetime.timedelta(days=days,
+                                   hours=self.hours,
+                                   minutes=self.minutes,
+                                   seconds=self.seconds))
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __rsub__(self, other):
+        return self.__neg__().__radd__(other)
+
+    def __sub__(self, other):
+        return NotImplemented
```

Comment:
I would argue that the subtraction operation needs to be commutative so that offset - date should make sense because it is just `offset - date = - (date - offset)`.  This is because the negation operator is defined and I'm using standard arithmetic set operations, which I think make sense in this context and is presumed by many.  Alternatively, we should issue a warning that this is not implemented, which is true in general for operators like this that seem like they should intuitively be implemented.